lib/install.ps1

```diff
@@ -666,88 +697,60 @@ function args($config, $dir, $global) {
-        Invoke-Expression (@($installer.script) -join "`r`n")
-        return
-    }
-
-    if($msi) {
-        install_msi $fname $dir $msi
-    } elseif($installer) {
-        install_prog $fname $dir $installer $global
-    }
-}
-
-# deprecated (see also msi_installed)
-function install_msi($fname, $dir, $msi) {
-    $msifile = "$dir\$(coalesce $msi.file "$fname")"
-    if(!(is_in_dir $dir $msifile)) {
-        abort "Error in manifest: MSI file $msifile is outside the app directory."
-    }
-    if(!($msi.code)) { abort "Error in manifest: Couldn't find MSI code."}
-    if(msi_installed $msi.code) { abort "The MSI package is already installed on this system." }
-
-    $logfile = "$dir\install.log"
-
-    $arg = @("/i `"$msifile`"", '/norestart', "/lvp `"$logfile`"", "TARGETDIR=`"$dir`"",
-        "INSTALLDIR=`"$dir`"") + @(args $msi.args $dir)
-
-    if($msi.silent) { $arg += '/qn', 'ALLUSERS=2', 'MSIINSTALLPERUSER=1' }
-    else { $arg += '/qb-!' }
-
-    $continue_exit_codes = @{ 3010 = "a restart is required to complete installation" }
+function Invoke-InstallerScript {
+    [CmdletBinding()]
+    param (
+        [String]
+        $AppName,
+        [String[]]
+        $FileName,
+        [PSObject]
+        $Manifest,
+        [String]
+        $Architecture,
+        [String]
+        $DestinationPath,
+        [Switch]
+        $Global
+    )
+    $Installer = installer $Manifest $Architecture
 
-    $installed = Invoke-ExternalCommand 'msiexec' $arg -Activity "Running installer..." -ContinueExitCodes $continue_exit_codes
-    if(!$installed) {
-        abort "Installation aborted. You might need to run 'scoop uninstall $app' before trying again."
+    if ($Installer.script) {
+        Write-Host "Running installer script..." -NoNewline
+        Invoke-Expression (@($Installer.script) -join "`r`n")
+        Write-Host "done." -f Green
+        return
     }
-    Remove-Item $logfile
-    Remove-Item $msifile
-}
-
-# deprecated
-# get-wmiobject win32_product is slow and checks integrity of each installed program,
-# so this uses the [wmi] type accelerator instead
-# http://blogs.technet.com/b/heyscriptingguy/archive/2011/12/14/use-powershell-to-find-and-uninstall-software.aspx
-function msi_installed($code) {
-    $path = "hklm:\software\microsoft\windows\currentversion\uninstall\$code"
-    if(!(test-path $path)) { return $false }
-    $key = Get-Item $path
-    $name = $key.getvalue('displayname')
-    $version = $key.getvalue('displayversion')
-    $classkey = "IdentifyingNumber=`"$code`",Name=`"$name`",Version=`"$version`""
-    try { $wmi = [wmi]"Win32_Product.$classkey"; $true } catch { $false }
-}
 
-function install_prog($fname, $dir, $installer, $global) {
-    $prog = "$dir\$(coalesce $installer.file "$fname")"
-    if(!(is_in_dir $dir $prog)) {
-        abort "Error in manifest: Installer $prog is outside the app directory."
+    if ($Installer.type) {
+        return
     }
```

Comment:
I wonder why?
Could the installer be excuted in such an order:
`installer.type -> installer.file -> installer.script` ?
Use case:
Calibre (not work because `installer.script` is excuted before `installer.file`): 

<details>
<summary>calibre.json</summary>

```
{
    "homepage": "https://calibre-ebook.com/",
    "version": "3.48.0",
    "description": "Powerful and easy to use e-book manager.",
    "license": "GPL-3.0-only",
    "url": "https://www.fosshub.com/Calibre.html/calibre-portable-installer-3.48.0.exe",
    "hash": "892ddb54caef1d4db065f7addf3e4d54e66141f3831c6740115cd43d9ddb430b",
    "persist": [
        "Calibre Library",
        "Calibre Settings"
    ],
    "installer": {
        "args": "\"$dir\"",
        "script": "Move-Item \"$dir\\Calibre Portable\\*\" \"$dir\""
    },
    "bin": "calibre-portable.exe",
    "shortcuts": [
        [
            "calibre-portable.exe",
            "Calibre"
        ]
    ],
    "checkver": {
        "url": "https://www.fosshub.com/Calibre.html",
        "regex": "softwareVersion\">([\\d.]+)"
    },
    "autoupdate": {
        "url": "https://www.fosshub.com/Calibre.html/calibre-portable-installer-$version.exe"
    }
}
```

</details>

Erlang (I test with this branch and remove above check, it works.):

<details>
<summary>erlang.json</summary>

```
{
    "homepage": "https://www.erlang.org",
    "description": "A programming language used to build massively scalable soft real-time systems with requirements on high availability.",
    "version": "22.0",
    "license": "Apache-2.0",
    "architecture": {
        "64bit": {
            "url": "http://erlang.org/download/otp_win64_22.0.exe#/dl.7z",
            "hash": "md5:f457e8aeda344a6743b2862feab16871"
        },
        "32bit": {
            "url": "http://erlang.org/download/otp_win32_22.0.exe#/dl.7z",
            "hash": "md5:2641d07ea5f6e723d56457a5d2ba82c2"
        }
    },
    "installer": {
        "type": "nsis",
        "file": "Install.exe",
        "args": "-sasl"
    },
    "bin": [
        "bin\\erl.exe",
        "bin\\erlc.exe",
        "bin\\escript.exe"
    ],
    "env_set": {
        "ERLANG_HOME": "$dir"
    },
    "shortcuts": [
        [
            "bin\\werl.exe",
            "Erlang"
        ]
    ],
    "post_install": "Remove-Item \"$dir\\vcredist*\" -Force",
    "checkver": {
        "url": "https://www.erlang.org/downloads",
        "regex": "DOWNLOAD\\s+OTP ([\\d.]+)"
    },
    "autoupdate": {
        "architecture": {
            "64bit": {
                "url": "http://erlang.org/download/otp_win64_$version.exe"
            },
            "32bit": {
                "url": "http://erlang.org/download/otp_win32_$version.exe"
            }
        },
        "hash": {
            "url": "$baseurl/MD5",
            "regex": "MD5\\($basename\\)= $md5"
        }
    },
    "suggest": {
        "Visual C/C++ Redistributable 2013": "extras/vcredist2013"
    }
}
```

</details>

Reply:
You could use `post_install` in such case... This seq should be discussed with @r15ch13 @rasa @Ash258 @h404bi etc.

Reply:
You cannot. As I think the point is it would break shortcuts creation,

Reply:
No, post_install doesn't work. In erlang case, I can use installer.script instead but installer.file is convient. In calibre case I can do nothing between installer.file and bin.

Reply:
So maybe move `installer.script` latest? installer.type -> installer.file -> installer.script?

Reply:
Yeah, that's what I want. It seems that there is no difference between pre_install and installer.script now.

Reply:
pre_install is mainly for use cases like renaming / moving installer.file and preparing needed files for installation. Like configuration files.

Reply:
Yes, just a small modification. `Calibre` is a good use case.

Reply:
Just try and test latest commit @linsui

Reply:
Not work for me

```
Installing 'calibre' (3.48.0) [64bit]
Loading calibre-portable-installer-3.48.0.exe from cache
Checking hash of calibre-portable-installer-3.48.0.exe ... ok.
Running installer... error.
ERROR Exception calling "Start" with "0" argument(s): "Access is denied"
Installation aborted. You might need to run 'scoop uninstall calibre' before trying again.
```